app/src/main/java/com/codefororlando/petadoption/feed/popular/PopularPetFeedPresenter.kt

```diff
@@ -0,0 +1,111 @@
+package com.codefororlando.petadoption.feed.popular
+
+import android.os.Bundle
+import com.codefororlando.petadoption.PetApplication
+import com.codefororlando.petadoption.data.model.Animal
+import com.codefororlando.petadoption.data.provider.IAnimalProvider
+import com.codefororlando.petadoption.feed.base.AbstractPetFeedPresenter
+import com.codefororlando.petadoption.recyclerview.AnimalListAdapter
+import io.reactivex.android.schedulers.AndroidSchedulers
+import io.reactivex.functions.Consumer
+import io.reactivex.schedulers.Schedulers
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+class PopularPetFeedPresenter : AbstractPetFeedPresenter<PopularPetFeedFragment>() {
+
+    private val ANIMAL_COUNT = 30
+    private val DEFAULT_OFFSET = "0"
+
+    @Inject
+    lateinit var animalProvider: IAnimalProvider
+
+    @Inject
+    lateinit var animalListAdapter: AnimalListAdapter
+
+    private var offset = DEFAULT_OFFSET
+
+    private var lastVisibleIndex = 0
+
+    override fun onCreate(savedState: Bundle?) {
+        super.onCreate(savedState)
+        PetApplication.getApp().appComponent()
+                .inject(this)
+    }
+
+    override fun onTakeView(petFeedFragment: PopularPetFeedFragment) {
+        super.onTakeView(petFeedFragment)
+        animalListAdapter.setOnItemClickListener { animal ->
+            view?.navigateToDetailView(animal)
+        }
+
+        petFeedFragment.setAdapter(animalListAdapter)
+        petFeedFragment.scrollToPosition(lastVisibleIndex)
+        compositeDisposable.add(view?.getOnScrollEndObservable()
```

Comment:
This can be simplified to

```kotlin
   view?.apply {
        compositeDisposable.add(getOnScrollEndObservable()
                .observeOn(AndroidSchedulers.mainThread())
                .debounce(300, TimeUnit.MILLISECONDS)
                .subscribe({this.loadMoreFeedItems()}, Timber::d))
   }
```

Remember that `?` ends up effectively creating null check code when the bytecode is generated so you want to use it sparingly and preferably never.